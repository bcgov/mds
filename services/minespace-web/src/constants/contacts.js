class Contact {
  constructor(title, name, phone, email) {
    this.title = title;
    this.name = name;
    this.phone = phone;
    this.email = email;
  }
}

class RegionalMineRegionalContacts {
  constructor(safety, permitting, director, office) {
    this.safety = safety;
    this.permitting = permitting;
    this.director = director;
    this.office = office;
  }
}

class MajorMineRegionalContacts {
  constructor(health) {
    this.health = health;
  }
}

const TITLE_SAFETY = "Senior Health, Safety and Environment Inspector";
const TITLE_PERMITTING = "Senior Permitting Inspector";
const TITLE_DIRECTOR = "Regional Director";
const TITLE_OFFICE = "Regional Office";

const REG_NE = new RegionalMineRegionalContacts(
  new Contact(TITLE_SAFETY, "Brian Oke", "(250) 565-4387", "Brian.Oke@gov.bc.ca"),
  new Contact(TITLE_PERMITTING, "Marnie Fraser", "(250) 565-4206", "Marnie.Fraser@gov.bc.ca"),
  new Contact(TITLE_DIRECTOR, "Victor Koyanagi", "(250) 565-4323", "Victor.Koyanagi@gov.bc.ca"),
  new Contact(TITLE_OFFICE, null, "(250) 565-4240", "MMD-PrinceGeorge@gov.bc.ca")
);

const REG_NW = new RegionalMineRegionalContacts(
  new Contact(TITLE_SAFETY, "Doug Flynn", "(250) 847-7386", "Doug.Flynn@gov.bc.ca"),
  new Contact(TITLE_PERMITTING, "Andrea Ross", "(250) 847-7768", "Andrea.Ross@gov.bc.ca"),
  new Contact(TITLE_DIRECTOR, "Howard Davies", "(250) 847-7653", "Howard.Davies@gov.bc.ca"),
  new Contact(TITLE_OFFICE, null, "(250) 847-7383", "MMD-Smithers@gov.bc.ca")
);

const REG_SC = new RegionalMineRegionalContacts(
  new Contact(TITLE_SAFETY, "Chris LeClair", "(250) 371-3714", "Chris.LeClair@gov.bc.ca"),
  new Contact(TITLE_PERMITTING, null, "(250) 371-3912", "MMD-Kamloops@gov.bc.ca"),
  new Contact(TITLE_DIRECTOR, "Rick Adams", "(250) 828-4583", "Rick.Adams@gov.bc.ca"),
  new Contact(TITLE_OFFICE, null, "(250) 371-3912", "MMD-Kamloops@gov.bc.ca")
);

const REG_SE = new RegionalMineRegionalContacts(
  new Contact(TITLE_SAFETY, "Michael Daigle", "(250) 417-6141", "Michael.Daigle@gov.bc.ca"),
  new Contact(TITLE_PERMITTING, "Glen Hendrickson", "(250) 417-6033", "Glen.Hendrickson@gov.bc.ca"),
  new Contact(TITLE_DIRECTOR, "Kathie Wagar", "(250) 417-6011", "Kathie.Wagar@gov.bc.ca"),
  new Contact(TITLE_OFFICE, null, "(250) 417-6134", "MMD-Cranbrook@gov.bc.ca")
);

const REG_SW = new RegionalMineRegionalContacts(
  new Contact(TITLE_SAFETY, "Jim Dunkley", "(778) 698-7294", "Jim.Dunkley@gov.bc.ca"),
  new Contact(TITLE_PERMITTING, "Don Harrison", "(778) 698-7014", "Donald.Harrison@gov.bc.ca"),
  new Contact(TITLE_DIRECTOR, "Matthew McLean", "(778) 698-9411", "Matthew.MacLean@gov.bc.ca"),
  new Contact(TITLE_OFFICE, null, "(778) 698-3649", "SouthwestMinesDivision@gov.bc.caâ€‹")
);

export const REGIONAL_MINE_REGIONAL_CONTACTS = {
  NE: REG_NE,
  NW: REG_NW,
  SC: REG_SC,
  SE: REG_SE,
  SW: REG_SW,
};

const TITLE_HEALTH = "Health and Safety Inspector";

const MAJ_NE = new MajorMineRegionalContacts(
  new Contact(TITLE_HEALTH, "Kristopher Bailey", "(250) 565-4271", "Kristopher.Bailey@gov.bc.ca")
);

const MAJ_NW = new MajorMineRegionalContacts(
  new Contact(TITLE_HEALTH, "Megan Frederick", "(250) 847-7214", "Megan.Frederick@gov.bc.ca")
);

const MAJ_SC = new MajorMineRegionalContacts(
  new Contact(TITLE_HEALTH, "Blythe Golobic", "(250) 371-3915", "Blythe.Golobic@gov.bc.ca")
);

const MAJ_SE = new MajorMineRegionalContacts(
  new Contact(TITLE_HEALTH, "Alan Day", "(250) 417-6013", "Alan.Day@gov.bc.ca")
);

const MAJ_SW = new MajorMineRegionalContacts(
  new Contact(TITLE_HEALTH, "Blythe Golobic", "(250) 371-3915", "Blythe.Golobic@gov.bc.ca")
);

export const MAJOR_MINE_REGIONAL_CONTACTS = {
  NE: MAJ_NE,
  NW: MAJ_NW,
  SC: MAJ_SC,
  SE: MAJ_SE,
  SW: MAJ_SW,
};

export const MM_OFFICE = new Contact("Major Mines Office", null, null, "PermRecl@gov.bc.ca");

export const CHIEF_INSPECTOR = new Contact(
  "Chief Inspector of Mines",
  "Hermanus Henning",
  "(778) 974-5980",
  "hermanus.henning@gov.bc.ca"
);

export const EXEC_LEAD_AUTH = new Contact(
  "Executive Lead (Authorizations)",
  "George Warnock",
  "(250) 649-4339",
  "brittanytownsend@gov.bc.ca"
);
