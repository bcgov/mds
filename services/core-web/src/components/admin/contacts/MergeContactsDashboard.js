import React, { Component } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { isEmpty } from "lodash";
import { Prompt } from "react-router-dom";
import { Row, Col, Card, Popconfirm, Button, Radio, Tabs, Alert } from "antd";
import { PhoneOutlined, MailOutlined } from "@ant-design/icons";
import PropTypes from "prop-types";
import RenderMultiSelectPartySearch from "@/components/common/RenderMultiSelectPartySearch";
import NullScreen from "@/components/common/NullScreen";
import {
  getPartyRelationshipTypeHash,
  getPartyRelationshipTypesList,
} from "@mds/common/redux/selectors/staticContentSelectors";
import { mergeParties } from "@mds/common/redux/actionCreators/partiesActionCreator";
import Address from "@/components/common/Address";
import * as Strings from "@mds/common/constants/strings";
import * as routes from "@/constants/routes";
import { openModal, closeModal } from "@mds/common/redux/actions/modalActions";
import { modalConfig } from "@/components/modalContent/config";
import * as Permission from "@mds/common/constants/permissions";
import AuthorizationWrapper from "@/components/common/wrappers/AuthorizationWrapper";
import { AuthorizationGuard } from "@/HOC/AuthorizationGuard";

const propTypes = {
  openModal: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  mergeParties: PropTypes.func.isRequired,
  partyRelationshipTypesHash: PropTypes.objectOf(PropTypes.string).isRequired,
  history: PropTypes.shape({ replace: PropTypes.func }).isRequired,
  location: PropTypes.shape({ pathname: PropTypes.string }).isRequired,
  match: PropTypes.shape({
    params: {
      tab: PropTypes.string,
    },
  }).isRequired,
};

// TODO: We can get these instead by using a static content selector?
const partyType = {
  PER: "Person",
  ORG: "Organization",
};

const codeFromURL = {
  Person: "PER",
  Organization: "ORG",
};

export class MergeContactsDashboard extends Component {
  state = {
    activeTab: "PER",
    expanded: false,
    selectedPartySearchResults: [],
    isLoading: false,
    contactsForMerge: [],
    rolesForMerge: [],
    triggerSelectReset: false,
    values: {
      first_name: "",
      party_name: "",
      email: "",
      phone_no: "",
      address: {},
    },
    valuesSelected: {
      first_name: "",
      party_name: "",
      email: "",
      phone_no: "",
      address: {},
    },
  };

  componentDidMount() {
    const { tab } = this.props.match.params;
    this.setState({
      activeTab: codeFromURL[tab],
    });
  }

  handleTabChange = (key) =>
    this.props.history.replace(routes.ADMIN_CONTACT_MANAGEMENT.dynamicRoute(partyType[key]));

  confirmMergeModal = () =>
    this.props.openModal({
      props: {
        title: "Confirm Merge",
        closeModal: this.props.closeModal,
        initialValues: this.state.values,
        onSubmit: this.handleMergeContacts,
        isPerson: this.state.activeTab === "PER",
        roles: this.state.rolesForMerge,
        partyRelationshipTypesHash: this.props.partyRelationshipTypesHash,
      },
      content: modalConfig.MERGE_PARTY_CONFIRMATION,
      width: "75vw",
    });

  setExpanded = () =>
    this.setState((prevState) => ({
      expanded: !prevState.expanded,
    }));

  handleClearState = () =>
    this.setState(
      {
        contactsForMerge: [],
        values: {},
        valuesSelected: {},
        rolesForMerge: [],
        selectedPartySearchResults: [],
        triggerSelectReset: true,
      },
      () => this.setState({ triggerSelectReset: false })
    );

  onSelectedPartySearchResultsChanged = (selectedPartySearchResults) =>
    this.setState({ selectedPartySearchResults });

  handleMergeContacts = (values) => {
    const payload = {
      party_guids: this.state.contactsForMerge.map(({ party_guid }) => party_guid),
      party: { ...values, party_type_code: this.state.activeTab },
    };
    this.setState({ isLoading: true });
    this.props
      .mergeParties(payload)
      .then(() => {
        this.handleClearState();
      })
      .finally(() => {
        this.props.closeModal();
        this.setState({ isLoading: false });
      });
  };

  shouldComponentUpdate = (nextProps, nextState) => {
    const searchSubsetChanged =
      this.state.selectedPartySearchResults !== nextState.selectedPartySearchResults;

    const tabChanged = this.props.match.params.tab !== nextProps.match.params.tab;
    if (tabChanged) {
      this.setState({
        activeTab: codeFromURL[nextProps.match.params.tab],
      });
      this.handleClearState();
    }

    if (searchSubsetChanged) {
      const contacts = [];
      const roles = [];
      // eslint-disable-next-line  array-callback-return, no-unused-expressions
      nextState.selectedPartySearchResults?.map((party) => {
        contacts.push(party);
        roles.push(...party?.mine_party_appt);
      });
      this.setState({ contactsForMerge: contacts, rolesForMerge: roles });
    }

    return true;
  };

  handleContactSelect = (event, field) => {
    this.setState((prevState) => ({
      values: { ...prevState.values, [field]: event.target.name },
      valuesSelected: { ...prevState.valuesSelected, [field]: event.target.value },
    }));
  };

  renderContactCards = () => {
    return (
      <>
        <div className="contact-container">
          <Alert
            description="All contacts selected will be deleted when the merge is complete and the proposed
          contact will be created."
            type="info"
            showIcon
            closable
          />
          <br />
          {this.state.contactsForMerge?.length > 0 ? (
            <Row gutter={[8, 8]} type="flex" justify="start">
              {this.state.contactsForMerge.map((data, i) => {
                return (
                  <Col span={6} flex={4} key={data.party_guid}>
                    <Card
                      className="no-header"
                      style={{ height: "100%", overflowX: "auto" }}
                      bordered={false}
                    >
                      {this.state.activeTab === "PER" && (
                        <Row>
                          <Col span={24} className="grid padding-sm">
                            <h6>First Name</h6>
                            <Radio.Group
                              name={data.first_name}
                              onChange={(event) => this.handleContactSelect(event, "first_name")}
                              value={this.state.valuesSelected.first_name}
                            >
                              <Radio value={`${data.first_name}${i}`} disabled={!data.first_name}>
                                {data.first_name || Strings.EMPTY_FIELD}
                              </Radio>
                            </Radio.Group>
                          </Col>
                        </Row>
                      )}
                      <Row>
                        <Col span={24} className="grid padding-sm">
                          <h6>
                            {this.state.activeTab === "PER" ? "Last Name" : "Organization Name"}
                          </h6>
                          <Radio.Group
                            name={data.party_name}
                            onChange={(event) => this.handleContactSelect(event, "party_name")}
                            value={this.state.valuesSelected.party_name}
                          >
                            <Radio value={`${data.party_name}${i}`} disabled={!data.party_name}>
                              {data.party_name || Strings.EMPTY_FIELD}
                            </Radio>
                          </Radio.Group>
                        </Col>
                      </Row>
                      <Row>
                        <Col span={24} className="grid padding-sm">
                          <h6>Phone Number</h6>
                          <Radio.Group
                            name={data.phone_no}
                            onChange={(event) => this.handleContactSelect(event, "phone_no")}
                            value={this.state.valuesSelected.phone_no}
                          >
                            <Radio value={`${data.phone_no}${i}`} disabled={!data.phone_no}>
                              {data.phone_no || Strings.EMPTY_FIELD}
                            </Radio>
                          </Radio.Group>
                        </Col>
                      </Row>
                      <Row>
                        <Col span={24} className="grid padding-sm">
                          <h6>Email</h6>
                          <Radio.Group
                            name={data.email}
                            onChange={(event) => this.handleContactSelect(event, "email")}
                            value={this.state.valuesSelected.email}
                          >
                            <Radio value={`${data.email}${i}`} disabled={!data.email}>
                              {data.email || Strings.EMPTY_FIELD}
                            </Radio>
                          </Radio.Group>
                        </Col>
                      </Row>
                      <Row>
                        <Col span={24} className="grid padding-sm">
                          <h6>Address</h6>
                          <Radio.Group
                            name={data.address[0]}
                            onChange={(event) => this.handleContactSelect(event, "address")}
                            value={this.state.valuesSelected.address}
                          >
                            <Radio value={data.address[0]} disabled={isEmpty(data.address[0])}>
                              <Address address={data.address[0] || {}} showIcon={false} />
                            </Radio>
                          </Radio.Group>
                        </Col>
                      </Row>
                      <Row>
                        <Col span={24} className="grid padding-sm">
                          <h6>Active Roles</h6>
                          {data?.mine_party_appt?.filter(({ end_date }) => !end_date)?.length >
                          0 ? (
                            data.mine_party_appt?.length > 0 &&
                            data.mine_party_appt
                              .filter(({ end_date }) => !end_date)
                              .map((appt) => (
                                <p>
                                  {
                                    this.props.partyRelationshipTypesHash[
                                      appt.mine_party_appt_type_code
                                    ]
                                  }{" "}
                                  - {appt.mine.mine_name}
                                </p>
                              ))
                          ) : (
                            <p>{Strings.EMPTY_FIELD}</p>
                          )}
                        </Col>
                      </Row>
                    </Card>
                  </Col>
                );
              })}
            </Row>
          ) : (
            <NullScreen type="merged-contacts" />
          )}
        </div>
      </>
    );
  };

  renderContactCard = (data) => (
    <Col span={24} key={data.party_guid}>
      <Card className="no-header inherit-height" bordered={false}>
        {this.state.activeTab === "PER" && (
          <Row>
            <Col span={24} className="grid padding-sm">
              <h6>First Name</h6>
              <p>{data.first_name || Strings.EMPTY_FIELD}</p>
            </Col>
          </Row>
        )}
        <Row>
          <Col span={24} className="grid padding-sm">
            <h6>{this.state.activeTab === "PER" ? "Last Name" : "Organization Name"}</h6>
            <p>{data.party_name || Strings.EMPTY_FIELD}</p>
          </Col>
        </Row>
        <Row>
          <Col span={24} className="grid padding-sm">
            <h6>Phone Number</h6>
            <p>
              <PhoneOutlined className="icon-sm padding-sm--right" />
              {data.phone_no || Strings.EMPTY_FIELD}
            </p>
          </Col>
        </Row>
        <Row>
          <Col span={24} className="grid padding-sm">
            <h6>Email</h6>
            <p>
              <MailOutlined className="icon-sm padding-sm--right" />
              {data.email || Strings.EMPTY_FIELD}
            </p>
          </Col>
        </Row>
        <Row>
          <Col span={24} className="grid padding-sm">
            <h6>Address</h6>
            <Address address={data.address || {}} />
          </Col>
        </Row>
        <div className="roles-container">
          <Row>
            <Col span={24} className="grid padding-sm">
              <h6>Active Roles</h6>
              <p>
                {this.state.rolesForMerge?.filter(({ end_date }) => !end_date)?.length > 0 ? (
                  this.state.rolesForMerge?.length > 0 &&
                  this.state.rolesForMerge
                    .filter(({ end_date }) => !end_date)
                    .map((appt) => (
                      <p>
                        {this.props.partyRelationshipTypesHash[appt.mine_party_appt_type_code]} -{" "}
                        {appt.mine.mine_name}
                      </p>
                    ))
                ) : (
                  <p>{Strings.EMPTY_FIELD}</p>
                )}
              </p>
            </Col>
          </Row>
          <Row>
            <Col span={24} className="grid padding-sm">
              <h6>In-active Roles</h6>
              <p>
                {this.state.rolesForMerge?.filter(({ end_date }) => end_date)?.length > 0 ? (
                  this.state.rolesForMerge?.length > 0 &&
                  this.state.rolesForMerge
                    .filter(({ end_date }) => end_date)
                    .map((appt) => (
                      <p>
                        {this.props.partyRelationshipTypesHash[appt.mine_party_appt_type_code]} -{" "}
                        {appt.mine.mine_name}
                      </p>
                    ))
                ) : (
                  <p>{Strings.EMPTY_FIELD}</p>
                )}
              </p>
            </Col>
          </Row>
        </div>
      </Card>
    </Col>
  );

  renderMergeContactsDashboard = () => (
    <>
      <div className="merge-dashboard">
        <h4 className="padding-sm">Merge {partyType[this.state.activeTab]}</h4>
        <Alert
          message="Not all contacts can be merged at this time"
          closable
          description={
            <>
              The following contacts will be disabled in the dropdown list:
              <div
                className={this.state.expanded ? "block" : "hidden"}
                style={{ marginLeft: "40px" }}
              >
                <ul>
                  <li>Contacts with the role of Permittee.</li>
                  <li>Contacts with the role of Inspector.</li>
                  <li>Organizations connected to OrgBook.</li>
                </ul>
              </div>
              <Button className="btn--expand" onClick={() => this.setExpanded()}>
                {this.state.expanded ? "  Read less" : "  ...Read more"}
              </Button>
            </>
          }
          type="info"
          showIcon
        />
        <br />
        <div className="search-contents inline-flex between">
          <div className="flex-1 padding-sm">
            <p>
              <b>Search and select contacts to merge</b>
            </p>
          </div>
          <div className="flex-4">
            <RenderMultiSelectPartySearch
              onSelectedPartySearchResultsChanged={this.onSelectedPartySearchResultsChanged}
              isLoading={this.state.isLoading}
              partyType={this.state.activeTab}
              triggerSelectReset={this.state.triggerSelectReset}
            />
          </div>
          <br />
        </div>
        <br />
        <div className="merge-container">
          <div className="merge-container--tall inline-flex between">
            <div className="flex-1">
              <h4 className="padding-sm">Proposed Merged Contact</h4>
              {this.renderContactCard(this.state.values)}
            </div>
            <div className="flex-4">{this.renderContactCards()}</div>
          </div>
          <br />
          <AuthorizationWrapper permission={Permission.ADMINISTRATIVE_USERS}>
            <div className="right center-mobile">
              <Popconfirm
                placement="topRight"
                title="Are you sure you want to cancel?"
                onConfirm={() => this.handleClearState()}
                okText="Yes"
                cancelText="No"
              >
                <Button className="full-mobile" type="secondary">
                  Clear All
                </Button>
              </Popconfirm>
              <Button
                className="full-mobile"
                type="primary"
                htmlType="submit"
                disabled={this.state.contactsForMerge?.length < 2}
                onClick={() => this.confirmMergeModal()}
              >
                Proceed to Merge
              </Button>
            </div>
          </AuthorizationWrapper>
        </div>
      </div>
    </>
  );

  render() {
    return (
      <div>
        <Prompt
          when={this.state.contactsForMerge.length > 0}
          message={(location) => {
            return this.props.location.pathname === location.pathname
              ? true
              : "Merge in progress. Are you sure you want to leave without saving? All progress will be lost.";
          }}
        />
        <div className="landing-page__header">
          <Row>
            <Col sm={22} md={14} lg={12}>
              <h1>Merge Contacts</h1>
            </Col>
          </Row>
          <Tabs
            activeKey={this.state.activeTab}
            size="large"
            animated={{ inkBar: false, tabPane: false }}
            onTabClick={this.handleTabChange}
            centered
          >
            <Tabs.TabPane tab="Merge Person" key="PER">
              <div className="tab__content">{this.renderMergeContactsDashboard()}</div>
            </Tabs.TabPane>
            <Tabs.TabPane tab="Merge Organization" key="ORG">
              <div className="tab__content">{this.renderMergeContactsDashboard()}</div>
            </Tabs.TabPane>
          </Tabs>
        </div>
      </div>
    );
  }
}

MergeContactsDashboard.propTypes = propTypes;

const mapStateToProps = (state) => ({
  partyRelationshipTypesList: getPartyRelationshipTypesList(state),
  partyRelationshipTypesHash: getPartyRelationshipTypeHash(state),
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      mergeParties,
      openModal,
      closeModal,
    },
    dispatch
  );

export default AuthorizationGuard(Permission.ADMINISTRATIVE_USERS)(
  connect(mapStateToProps, mapDispatchToProps)(MergeContactsDashboard)
);
