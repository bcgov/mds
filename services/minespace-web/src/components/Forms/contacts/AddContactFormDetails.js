import React, { useEffect, useState } from "react";
import {
  createParty,
  updateParty,
  fetchParties,
} from "@mds/common/redux/actionCreators/partiesActionCreator";
import { getParties } from "@mds/common/redux/selectors/partiesSelectors";
import { compose, bindActionCreators } from "redux";
import { Field, reduxForm, initialize, isDirty, reset, getFormValues, change } from "redux-form";
import { connect } from "react-redux";
import { Col, Row, Typography, Popconfirm, Button, Divider } from "antd";
import Form from "@ant-design/compatible/Form";
import { debounce } from "lodash";
import { getPartyRelationshipTypesList } from "@mds/common/redux/selectors/staticContentSelectors";

import {
  required,
  email,
  phoneNumber,
  validateSelectOptions,
  maxLength,
} from "@common/utils/Validate";
import { normalizePhone } from "@common/utils/helpers";
import PropTypes from "prop-types";
import { renderConfig } from "@/components/common/config";
import * as FORM from "@/constants/forms";
import { party as PartyPropType, partyRelationshipType } from "@/customPropTypes/parties";

const propTypes = {
  createParty: PropTypes.func.isRequired,
  updateParty: PropTypes.func.isRequired,
  fetchParties: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  handleSelectChange: PropTypes.func.isRequired,
  change: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  isDirty: PropTypes.bool.isRequired,
  formValues: PropTypes.objectOf(PartyPropType).isRequired,
  organizations: PropTypes.arrayOf(PartyPropType).isRequired,
  contacts: PropTypes.arrayOf(PartyPropType),
  partyRelationshipTypesList: PropTypes.arrayOf(partyRelationshipType).isRequired,

  // eslint-disable-next-line react/no-unused-prop-types
  initialValues: PropTypes.objectOf(PartyPropType).isRequired,
};

const defaultProps = {
  contacts: [],
};

export const AddContactFormDetails = (props) => {
  const [submitting, setSubmitting] = useState(false);

  const onSubmit = async (values) => {
    const party_type_code = "PER";
    const payload = { party_type_code, ...values };

    setSubmitting(true);

    try {
      if (!values.party_guid) {
        // Party doesn't already exist, create it
        const { data: party } = await props.createParty(payload);

        await props.onSubmit(party);
      } else if (props.isDirty) {
        // Selected party has been updated, update it
        const { data: party } = await props.updateParty(payload, values.party_guid);

        await props.onSubmit(party);
      } else {
        // Selected party has not been updated, use it as is
        await props.onSubmit(values);
      }
    } finally {
      setSubmitting(true);
    }
  };

  const getSubmitText = () => {
    if (!props.formValues?.party_guid) {
      return "Create Contact";
    }

    if (props.isDirty) {
      return "Update and Select";
    }

    return "Select Contact";
  };

  const searchOrganizations = (search) => {
    props.fetchParties({
      type: "ORG",
      party_name: search,
      per_page: 10,
    });
  };

  useEffect(() => {
    searchOrganizations("");
  }, []);

  const transformOrganizations = (orgs) =>
    Object.values(orgs).map((org) => ({
      label: org.name,
      value: org.party_guid,
    }));

  const handleSelectChange = (e, val, oldVal, field) => {
    // Sets the value of the given select field to `val`
    // defaults a missing value to `null` instead of `undefined`
    // which allows the select component to clear the existing value instead of
    // defaulting to the initial value when "clear" is clicked
    props.change(FORM.ADD_CONTACT, field, val || null);
  };

  return (
    <Form layout="vertical" onSubmit={props.handleSubmit(onSubmit)}>
      <Row gutter={16}>
        <Col span={24}>
          <Typography.Paragraph>
            Create a new contact for your organization. New contacts will be accessible accross all
            of your organizations mines during submission.
          </Typography.Paragraph>
        </Col>
        <Col span={12}>
          <Form.Item label="Select Contact">
            <Field
              id="party_guid"
              name="party_guid"
              placeholder="Select a contact"
              component={renderConfig.SELECT}
              onChange={props.handleSelectChange}
              data={props.contacts}
              validate={[validateSelectOptions(props.contacts, true)]}
            />
          </Form.Item>
        </Col>
        <Col span={24}>
          <Divider plain style={{ marginTop: 0 }} />
        </Col>
      </Row>
      <Row gutter={16}>
        <Col span={24}>
          <Typography.Title level={5}>Contact Details</Typography.Title>

          <Typography.Paragraph>
            {props.formValues?.party_guid
              ? "If this contact requires edit before selection you can do so below."
              : "Please enter all contact information for this contact."}
          </Typography.Paragraph>
        </Col>
        <Col span={12}>
          <Form.Item label="First Name">
            <Field
              id="first_name"
              name="first_name"
              placeholder="First Name"
              component={renderConfig.FIELD}
              validate={[required, maxLength(200)]}
            />
          </Form.Item>
        </Col>
        <Col span={12}>
          <Form.Item label="Last Name">
            <Field
              id="party_name"
              name="party_name"
              placeholder="Last Name"
              component={renderConfig.FIELD}
              validate={[required, maxLength(200)]}
            />
          </Form.Item>
        </Col>
        <Col span={12}>
          <Form.Item label="Job Title (optional)">
            <Field
              id="job_title_code"
              name="job_title_code"
              placeholder="Select a job title"
              onChange={handleSelectChange}
              component={renderConfig.SELECT}
              data={props.partyRelationshipTypesList}
              validate={[validateSelectOptions(props.partyRelationshipTypesList)]}
            />
          </Form.Item>
        </Col>
        <Col span={12}>
          <Form.Item label="Company Affiliation (optional)">
            <Field
              id="organization_guid"
              name="organization_guid"
              onChange={handleSelectChange}
              component={renderConfig.AUTOCOMPLETE}
              placeholder="Search organizations"
              data={transformOrganizations(props.organizations)}
              handleChange={searchOrganizations}
            />
          </Form.Item>
        </Col>

        <Col span={12}>
          <Form.Item label="Email">
            <Field
              id="email"
              name="email"
              component={renderConfig.FIELD}
              placeholder="example@example.com"
              validate={[email, required]}
            />
          </Form.Item>
        </Col>
        <Col span={8}>
          <Form.Item label="Phone Number">
            <Field
              name="phone_no"
              id="phone_no"
              placeholder="XXX-XXX-XXXX"
              component={renderConfig.FIELD}
              validate={[required, phoneNumber, maxLength(12)]}
              normalize={normalizePhone}
            />
          </Form.Item>
        </Col>
        <Col span={4}>
          <Form.Item label="Ext. (optional)">
            <Field
              name="phone_ext"
              id="phone_ext"
              component={renderConfig.FIELD}
              validate={[maxLength(6)]}
            />
          </Form.Item>
        </Col>
      </Row>
      <Row justify="space-between">
        <Popconfirm
          placement="top"
          title="Are you sure you want to cancel?"
          okText="Yes"
          cancelText="No"
          onConfirm={props.onCancel}
        >
          <Button disabled={submitting} className="full-mobile">
            Cancel
          </Button>
        </Popconfirm>
        <Button disabled={submitting} type="primary" className="full-mobile" htmlType="submit">
          {getSubmitText()}
        </Button>
      </Row>
    </Form>
  );
};

AddContactFormDetails.propTypes = propTypes;
AddContactFormDetails.defaultProps = defaultProps;

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      createParty,
      updateParty,
      fetchParties: (...args) => debounce(() => dispatch(fetchParties(...args)), 1000),
      initialize: (data) => initialize(FORM.ADD_CONTACT, data),
      reset,
      change,
    },
    dispatch
  );

const mapStateToProps = (state) => ({
  isDirty: isDirty(FORM.ADD_CONTACT)(state),
  formValues: getFormValues(FORM.ADD_CONTACT)(state),
  partyRelationshipTypesList: getPartyRelationshipTypesList(state),
  organizations: getParties(state),
});

export default compose(
  reduxForm({
    form: FORM.ADD_CONTACT,
    destroyOnUnmount: true,
    forceUnregisterOnUnmount: true,
    enableReinitialize: true,
    touchOnBlur: true,
  }),
  connect(mapStateToProps, mapDispatchToProps)
)(AddContactFormDetails);
