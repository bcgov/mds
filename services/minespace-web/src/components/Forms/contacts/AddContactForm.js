import React, { useState } from "react";
import { connect } from "react-redux";
import { uniqBy } from "lodash";
import { getPartyRelationships } from "@mds/common/redux/selectors/partiesSelectors";
import PropTypes from "prop-types";
import AddContactFormDetails from "./AddContactFormDetails";

import { partyRelationship as PartyRelationShipType } from "@/customPropTypes/parties";

const propTypes = {
  parties: PropTypes.arrayOf(PartyRelationShipType).isRequired,
  mine_party_appt_type_code: PropTypes.string.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
};

const defaultProps = {};

export const AddContactForm = (props) => {
  const [selectedParty, setSelectedParty] = useState(null);

  const handleSelectChange = (evt, val) => {
    if (!val) {
      setSelectedParty(null);
    } else {
      const partyRelationship = props.parties.find((p) => p.party.party_guid === val);

      if (partyRelationship) {
        setSelectedParty(partyRelationship.party);
      }
    }
  };

  const contacts = uniqBy(props.parties || [], (partyAppt) => partyAppt.party.party_guid)
    .filter(
      ({ mine_party_appt_type_code }) =>
        !props.mine_party_appt_type_code ||
        mine_party_appt_type_code === props.mine_party_appt_type_code
    )
    .map(({ party: { name, party_guid } }) => ({
      label: name,
      value: party_guid,
    }));

  return (
    <AddContactFormDetails
      contacts={contacts}
      initialValues={selectedParty}
      handleSelectChange={handleSelectChange}
      onCancel={props.onCancel}
      onSubmit={props.onSubmit}
    />
  );
};

AddContactForm.propTypes = propTypes;
AddContactForm.defaultProps = defaultProps;

const mapStateToProps = (state) => ({
  parties: getPartyRelationships(state),
});

export default connect(mapStateToProps)(AddContactForm);
