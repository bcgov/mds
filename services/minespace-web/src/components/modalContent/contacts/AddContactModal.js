import React from "react";
import PropTypes from "prop-types";
import AddContactForm from "@/components/Forms/contacts/AddContactForm";

const propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
};

const AddContactModal = (props) => {
  const { onSubmit, onCancel } = props;

  return <AddContactForm onSubmit={onSubmit} onCancel={onCancel} />;
};

AddContactModal.propTypes = propTypes;

export default AddContactModal;
