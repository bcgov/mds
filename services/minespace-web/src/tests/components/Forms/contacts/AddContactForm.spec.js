import React from "react";
import { shallow } from "enzyme";
import { AddContactForm } from "@/components/Forms/contacts/AddContactForm";

const props = {};
const dispatchProps = {};

const setupDispatchProps = () => {
  dispatchProps.onCancel = jest.fn();
  dispatchProps.onSubmit = jest.fn();
};

const setupProps = () => {
  props.contacts = [];
  props.parties = [];
};

beforeEach(() => {
  setupDispatchProps();
  setupProps();
});

describe("AddContactForm", () => {
  it("renders properly", () => {
    const component = shallow(<AddContactForm {...props} {...dispatchProps} />);
    expect(component).toMatchSnapshot();
  });
});
