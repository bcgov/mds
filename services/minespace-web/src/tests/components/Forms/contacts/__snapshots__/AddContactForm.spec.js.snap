// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddContactForm renders properly 1`] = `
<ReduxForm
  contacts={Array []}
  destroyOnUnmount={true}
  enableReinitialize={true}
  forceUnregisterOnUnmount={true}
  form="ADD_CONTACT"
  getFormState={[Function]}
  handleSelectChange={[Function]}
  initialValues={null}
  keepDirtyOnReinitialize={false}
  onCancel={[MockFunction]}
  onSubmit={[MockFunction]}
  parties={Array []}
  persistentSubmitErrors={false}
  pure={true}
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  submitAsSideEffect={false}
  touchOnBlur={true}
  touchOnChange={false}
  updateUnregisteredFields={false}
/>
`;
